compiler/src/dotty/tools/backend/sjs/JSEncoding.scala

```diff
@@ -398,8 +398,17 @@ object JSEncoding {
     paramAndResultTypeNames.mkString(SignatureSep, SignatureSep, "")
 
   /** Computes the internal name for a type. */
-  private def internalName(tpe: Type)(implicit ctx: Context): String =
-    encodeTypeRef(toTypeRef(tpe))
+  private def internalName(tpe: Type)(implicit ctx: Context): String = {
+    val typeRef = toTypeRef(tpe)
+
+    val safeTypeRef: jstpe.TypeRef = typeRef match {
+      case jstpe.ClassRef("sr_Null" | "s_Null")    => jstpe.ClassRef(ir.Definitions.NullClass)
+      case jstpe.ClassRef("sr_Nothing" | "s_Nothing") => jstpe.ClassRef(ir.Definitions.NothingClass)
```

Comment:
Hum ... normally you should exclusively test for `sr_Null$` and `sr_Nothing$` at this point, similarly to what's done for the scalac side at
https://github.com/scala-js/scala-js/blob/68bae6f971f2f8176d1d36937d1e5f07c4fadeaf/compiler/src/main/scala/org/scalajs/nscplugin/JSEncoding.scala#L285-L288

I'm surprised the test even passes if you're matching `sr_Null` (without `$`) and `s_Null`. Neither should even happen at this point :s

Reply:
Investigation so far: Scala 2's version uses a `needsModuleClassSuffix` to add the `$` in `encodeClassFullName`, where Scala 3's version does not do so directly (so far as i can tell, `fullyMangledString`/`mangledName` encodes characters in the `NameTransformer`, but does not add the `$` suffix, which may be more the domain of `.fullName` here).

It appears that in Scala 3 the `ModuleClassName` `SuffixNameKind` is responsible for adding the `$` suffix, though I'm still working through how that works/where that happens.

Reply:
OK, thanks for thé investigation.

I'm not sure sure yet why this happens the way it happens, but given the investigation and that the test passes, I propose we book this progress as is.

Could you just check which one of `s_Null` or `sr_Null` is actually required to make the test pass, and drop the other alternative?

Reply:
Done - surprisingly, it is the the `s_Null` that is coming through, not the `sr_Null`, which hints that whatever usually maps the language version to the runtime version isn't getting used here. Something to dig into later.